File: 2022/day_04/s.jl
function y_contains_x(x, y)
   return x[1] >= y[1] && x[2] <= y[2]
end

function part1(filename)
  data = read(filename, String) |> txt -> split(strip(txt), "\n") .|> pairs->map(ranges->map(elt->parse(Int, elt), split(ranges,"-")),split(pairs, ","))
  res = sum(map(x->y_contains_x(x[1],x[2]) || y_contains_x(x[2],x[1]), data))
  return res
end


function overlaps(x,y)
  return (y_contains_x(x,y) || y_contains_x(y,x) || x[2] >= y[1] || y[2] >= x[1]) && !(x[2] < y[1] || y[2] < x[1]) 
end

function part2(filename)
  data = read(filename, String) |> (txt -> split(strip(txt), "\n")) .|> line -> split(line,",") .|> (pair -> split(pair, "-") .|> (x->parse(Int,x)))
  res = sum(map(x->overlaps(x[1],x[2]), data))
  return res
end

filename = "example.txt"
println(part1(filename))
println(part2(filename))

File: 2025/day_11/s.py
from collections import defaultdict
from functools import cache
from pathlib import Path

example = Path(__file__).parent / "example.txt"
example2 = Path(__file__).parent / "example2.txt"
true_input = Path(__file__).parent / "input.txt"
to_read = true_input

print(f"running with {to_read.name}")
data = open(to_read).read().strip().split("\n")


def parse(data):
    graph = defaultdict(list)

    for line in data:
        source, dests = line.split(": ")
        dests = dests.split(" ")
        graph[source].extend(dests)

    return graph


def dfs(g, src, dst):
    search = [src]
    found_paths = 0

    while len(search) > 0:
        pos = search.pop()

        for neighbor in g[pos]:
            if neighbor == dst:
                found_paths += 1
                continue
            search.append(neighbor)

    return found_paths


def part1(data: list[str]):
    g = parse(data)
    found = dfs(g, "you", "out")
    return found


g = parse(data)


@cache
def num_paths(src, dst, wentbyfft=False, wentbydac=False):
    if src == dst:
        return 0
    if dst in g[src]:
        if wentbyfft and wentbydac:
            return 1
        else:
            return 0
    sources_to_dest = []
    for gsrc, gdsts in g.items():
        if dst in gdsts:
            sources_to_dest.append(gsrc)
    return sum(
        [
            num_paths(
                src,
                new_dst,
                wentbyfft or new_dst == "fft",
                wentbydac or new_dst == "dac",
            )
            for new_dst in sources_to_dest
        ]
    )


def part2(data: list[str]):
    return num_paths("svr", "out")


print(part1(data))
print(part2(data))
